Common storage backend config

colega opened the pull request:
#### What this PR does

Adds a `common` configuration block to yaml, as well as `common.`-prefixed flags that will allow configuring common values for all services.

In this first iteration, we'll allow configuring a common storage backend.

**The logic this follows is: flags override yaml, specific configs override common.**

In particular, this means that a specific yaml configuration will be overridden by a common flag value (there's a test case for that). This is not exacty what I desired (I'd prefer to have "specific over common, flags over yaml" instead), but given the complexity of implementing the other option, I think we can accept the tradeoff.

_Why is it complex to do the other option?_

We would need to know how flags relate to specific struct fields (reflection-based) in order to know whether a common flag name doesn't have to be applied because it was set by yaml.

This is already hard since there's no \[decent\] way of sniffing what fields are being registered in the flagset with which names (actually, there might be no strict 1-1 relation between flags and fields, since `flag.Value` is just an interface that does _stuff_).

If we were able to achieve that, we'd still need to be able to distinguish whether a specific config value is there just because it was a default value or otherwise it's because yaml stated the same value as the default one.

#### Which issue(s) this PR fixes or relates to

Relates to https://github.com/grafana/mimir/issues/2319

#### Checklist

- [x] Tests updated
- [ ] Documentation added: https://github.com/grafana/mimir/pull/2347
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented on pkg/mimir/mimir.go:
```diff
@@ -158,6 +160,70 @@ func (c *Config) RegisterFlags(f *flag.FlagSet, logger log.Logger) {
+		return fmt.Errorf("can't unmarshal common config: %w", err)
+	}
+
+	// Then unmarshal config in a standard way.
```
[nit] I would extend the comment to clarify that if the specific locations config options are set, this will override the common settings that were previously set, so that "YAML specific" takes precedence over "YAML common".

colega replied:
Clarified in https://github.com/grafana/mimir/pull/2330/commits/e1a9253e4aba10141b8a8c1522cf190980fa5c2f

pracucci reviewed (approved):
Solid work! LGTM.

We need documentation for that, but it can be done in a follow up PR, so that we don't block it on doc reviews.

colega commented:
I've updated some tests and the tutorial to use common flags (I've left couple of tests using specific configurations, to have both of them).

I also extracted root blocks for common and providers configs as you suggested, @pracucci. 

Re-requesting review since there have been a bunch of changes here, meanwhile I'll start working on adding some better docs for this (I'll include docs squad as reviewer in those).

pracucci reviewed (approved):
LGTM! (modulo few minor comments)

Have you manually tried the tutorial? If yes, LGTM, otherwise please test it 😉

colega commented on docs/sources/operators-guide/configuring/reference-configuration-parameters/index.md:
```diff
@@ -3928,3 +3406,276 @@ The `memcached` block configures the Memcached-based caching backend. The suppor
+
+# The sse block configures the S3 server-side encryption.
+# The CLI flags prefix for this block configuration is: common.storage
+[sse: <sse>]
```
Oh good point, removed in https://github.com/grafana/mimir/pull/2330/commits/f614052b6b7f8329a0c3c99e52033c73c12d1a6e

colega commented on docs/sources/operators-guide/configuring/reference-configuration-parameters/index.md:
```diff
@@ -214,6 +214,47 @@ runtime_config:
+  # The s3_backend block configures the connection to Amazon S3 object storage
+  # backend.
+  # The CLI flags prefix for this block configuration is: common.storage
+  [s3: <s3_storage_backend>]
```
Yes, I followed the same reasoning, wanted to go with just `s3`, `gcs`, etc, but they just looked too generic here, especially `filesystem`, which is not an object-storage per se.

pracucci commented:
LGTM merging it.

dbrennand commented:
Hi @colega @pracucci 

Is there a container image available to use these changes?

The most recently updated image tag available on DockerHub is [r193-df12d7b](https://hub.docker.com/layers/mimir/grafana/mimir/r193-df12d7b/images/sha256-526cefa9ddc86b6072e32ec5f441a8a1222957e3083afec5a701a1a8ed07b268?context=explore) which I don't think includes these changes yet:

```
error loading config from /etc/mimir/mimir.yml: Error parsing config file: yaml: unmarshal errors:
  line 10: field common not found in type mimir.Config
```

replay commented:
This is a great feature, thanks a lot for implementing it!
Do we need to update `backend-enterprise` so that the GEM specific block storage config also uses the config from `common`?

pracucci commented:
> This is a great feature, thanks a lot for implementing it! Do we need to update `backend-enterprise` so that the GEM specific block storage config also uses the config from `common`?

Already done.

colega commented:
Now we should start using it 💪
